[DS-36] 사장님 Repository 추가

rnjstmdals6 opened the pull request:
## Goals
- 사장님 Entity 를 저장 및 구성을 위한 BossRepository 를 추가합니다.
- QueryDsl 기능을 사용하기 위해 BossQueryRepository 를 추가합니다
- 중복 가입 방지를 위해 이메일, 가게 등록 번호, 핸드폰 번호가 저장소에 존재하는 지 체크하는 매서드를 추가합니다.

## Implements
- [x] 중복 가입 방지 JPA 매서드 구현하고 단위 테스트를 진행합니다.

## Related Issue 
- https://dangol-sonnim.atlassian.net/browse/DS-36?atlOrigin=eyJpIjoiNjVjMjdiOTEzNjI5NDU3YmFmNmZjNmRmODk5NmU5ZjkiLCJwIjoiaiJ9

## Notes
해당 브랜치는 [feature/DS-35-add-boss-entity] 브랜치를 기반으로 작성되었습니다.

rnjstmdals6 committed:
feat: BossRepository 추가

Boss 개체의 CRUD를 담당합니다.

rnjstmdals6 committed:
feat: BossQueryRepository 추가

해당 DB에
이메일 존재 여부
핸드폰 번호 존재 여부
사업자 등록 번호 존재 여부를 체크합니다.

rnjstmdals6 committed:
Revert "feat: BossQueryRepository 추가"

This reverts commit 62290cd77e18af6fdc43a2feda514a886e97e0c3.

rnjstmdals6 committed:
Revert "Revert "feat: BossQueryRepository 추가""

This reverts commit e72dd129c38c64a6c0e273ae40c644b2685fc66d.

rnjstmdals6 committed:
test: BossRepository 저장 테스트 추가

rnjstmdals6 commented:
로컬에서는 테스트가 정상적으로 통과가 되는 데
깃 액션 빌드 시 DB 커넥션 오류가 납니다

y00njinuk committed:
feat: add @AutoConfigureTestDatabase annotation for test

[issue-key] DS-36
- 테스트 시 H2 데이터베이스를 사용하지 않고 MySQL 데이터베이스를 사용하도록 설정
- Reference: 스프링 부트 핵심가이드 p.192~193

y00njinuk commented on src/test/java/com/dangol/dangolsonnimbackend/boss/repository/BossRepositoryTest.java:
```diff
@@ -5,13 +5,16 @@
 import static org.junit.jupiter.api.Assertions.*;
 
 @DataJpaTest
+@AutoConfigureTestDatabase(replace = Replace.NONE)
```
> 로컬에서는 테스트가 정상적으로 통과가 되는 데 깃 액션 빌드 시 DB 커넥션 오류가 납니다

DataJpa 어노테이션 사용 시 임베디드 된 H2 데이터베이스를 사용하기 때문에 위와 같은 annotation을 추가하여 실제 MySQL 데이터베이스를 사용할 수 있게끔 해야한다고 하네요. (스프링 부트 핵심가이드 p.192~193 참조)

rnjstmdals6 replied:
감사합니다!!

y00njinuk commented on src/test/java/com/dangol/dangolsonnimbackend/boss/repository/BossRepositoryTest.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Test
+    void whenFindById_thenReturnBoss() {
```
given 표시가 없네요?

rnjstmdals6 replied:
여기 given 은 dto에 해당됩니다.
하지만 dto를 생성하는데 중복코드가 발생하기 때문에 setup 매서드로 따로 분리하였습니다.

이미 설정 방법에서 시스템의 초기 상태가 설정되고 있음이 명확하고 
해당 메서드는 수행 중인 작업과 예상 결과에 맞춰져 있기 때문에 "when"으로 시작하여 "then"으로 네이밍 하였습니다.

y00njinuk commented on src/test/java/com/dangol/dangolsonnimbackend/boss/repository/dsl/BossQueryRepositoryTest.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Test
+    public void givenValidEmail_whenExistsByEmail_thenReturnTrue() {
```
여기는 given 표시가 있던데, 위에서 진행한 BossRepositoryTest에 표기에는 없네요.

rnjstmdals6 replied:
해당 테스트코드는 메서드 내에서 Email 이라는 데이터가 주어지지만

위에서 진행한 BossRepositoryTest에서 존재하는 테스트 코드들의 given 은 setup 매서드 내에서 명시하기 때문입니다.

y00njinuk replied:
BossRepositoryTest 는 JPA에서 제공하는 기본적인 API를 활용한 테스트이고
BossQueryRepositoryTest는 QueryDSL을 통해 직접 정의한 API를 활용한 테스트라고 이해해도 될까요?

rnjstmdals6 replied:
넵!

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/boss/repository/dsl/BossQueryRepository.java:
```diff
@@ -0,0 +1,40 @@
+    private final JPAQueryFactory queryFactory;
+
+    public boolean existsByEmail(String email) {
+        Integer fetchOne = queryFactory.selectOne()
```
아하... QueryDSL은 이런 방식으로 쿼리를 정의하는군요.

rnjstmdals6 replied:
금방 배우실 수 있으실 겁니다!

Hyeonziii commented on src/main/java/com/dangol/dangolsonnimbackend/boss/repository/dsl/BossQueryRepository.java:
```diff
@@ -0,0 +1,40 @@
+        return fetchOne != null;
+    }
+
+    public boolean existsBySrn(String srn) {
```
String storeRegisterNumber로 변경 부탁드려요!

rnjstmdals6 replied:
수정 완료 했습니다 감사합니다!

Hyeonziii commented on src/test/java/com/dangol/dangolsonnimbackend/boss/repository/BossRepositoryTest.java:
```diff
@@ -0,0 +1,54 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@DataJpaTest
```
SpringBootTest로 변경 부탁드려요!

rnjstmdals6 replied:
수정했습니다!

y00njinuk reviewed (changes requested):
https://github.com/jaemisam-a/dangol-sonnim-backend/pull/8/files#diff-12847f3a2796d886d6d98904b6a903252b0de84310833b9a56ac27a088b30bb0R16 

SpringBootTest로 어노테이션 변경 부탁드립니다.

rnjstmdals6 committed:
feat: @DataJpaTest 에서 @SpringBootTest 로 전환

단위 테스트가 아닌 통합 테스트로 해당 클래스에 대한 테스트를 진행합니다. 여기서 @Transactional 이란 메소드 내에서 데이터베이스에 대한 변경사항은 메소드가 완료된 후 롤백되어 데이터베이스가 원래 상태로 유지됩니다. 각 테스트 케이스가 다른 테스트 케이스와 독립적으로 실행되고 데이터베이스 상태가 다른 테스트 케이스보다 먼저 실행되는 테스트 케이스의 영향을 받지 않도록 하기 위해 사용했습니다!
